Kernel: We need to use the Python 3.9.7 from the Anaconda distribution.

#  **1. Import libraries**

First, we need to import some libraries for data processing.

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

* `pandas:` offers data structures and operations to manipulate numerical tables and time series.
* `numpy:` offers a giant collection of high-level mathematical functions to operate on arrays and matrices.
* `matplotlib:` offers an interactive visualization lile plotting graphs.

# **2. Data Collection**

## ***2.1 Get raw data***

Our data is stored in a CSV file.

In [4]:
data_dir = '../Data/USW00013984.csv'
df = pd.read_csv(data_dir)
df

STATION  DATE  LATITUDE  LONGITUDE  ELEVATION                   NAME  \
0    USW00013984  1886  39.55127  -97.65077      444.3  CONCORDIA ASOS, KS US   
1    USW00013984  1887  39.55127  -97.65077      444.3  CONCORDIA ASOS, KS US   
2    USW00013984  1888  39.55127  -97.65077      444.3  CONCORDIA ASOS, KS US   
3    USW00013984  1889  39.55127  -97.65077      444.3  CONCORDIA ASOS, KS US   
4    USW00013984  1890  39.55127  -97.65077      444.3  CONCORDIA ASOS, KS US   
..           ...   ...       ...        ...        ...                    ...   
119  USW00013984  2018  39.55127  -97.65077      444.3  CONCORDIA ASOS, KS US   
120  USW00013984  2019  39.55127  -97.65077      444.3  CONCORDIA ASOS, KS US   
121  USW00013984  2020  39.55127  -97.65077      444.3  CONCORDIA ASOS, KS US   
122  USW00013984  2021  39.55127  -97.65077      444.3  CONCORDIA ASOS, KS US   
123  USW00013984  2022  39.55127  -97.65077      444.3  CONCORDIA ASOS, KS US   

     AWND AWND_ATTRIBUTES   CDSD  CDSD_ATTRIBUTES  ...  WSF1 WSF1_ATTRIBUTES  \
0     NaN             NaN  741.6              NaN  ...   NaN             NaN   
1     NaN             NaN  710.0              NaN  ...   NaN             NaN   
2     NaN             NaN  644.0              NaN  ...   NaN             NaN   
3     NaN             NaN  535.0              NaN  ...   NaN             NaN   
4     NaN             NaN  743.1              NaN  ...   NaN             NaN   
..    ...             ...    ...              ...  ...   ...             ...   
119   4.6               W  829.7              NaN  ...   NaN             NaN   
120   4.7               W  899.9              NaN  ...   NaN             NaN   
121   4.6               W  937.3              NaN  ...   NaN             NaN   
122   4.7               W    NaN              NaN  ...   NaN             NaN   
123   NaN             NaN    NaN              NaN  ...   NaN             NaN   

     WSF2 WSF2_ATTRIBUTES  WSF5 WSF5_ATTRIBUTES  WSFG WSFG_ATTRIBUTES  WSFM  \
0     NaN             NaN   NaN             NaN   NaN             NaN   NaN   
1     NaN             NaN   NaN             NaN   NaN             NaN   NaN   
2     NaN             NaN   NaN             NaN   NaN             NaN   NaN   
3     NaN             NaN   NaN             NaN   NaN             NaN   NaN   
4     NaN             NaN   NaN             NaN   NaN             NaN   NaN   
..    ...             ...   ...             ...   ...             ...   ...   
119  18.3               W  26.8               W   NaN             NaN   NaN   
120  17.9               W  27.3               W   NaN             NaN   NaN   
121  19.7               W  28.2               W   NaN             NaN   NaN   
122  24.6               W  39.3               W   NaN             NaN   NaN   
123   NaN             NaN   NaN             NaN   NaN             NaN   NaN   

     WSFM_ATTRIBUTES  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
4                NaN  
..               ...  
119              NaN  
120              NaN  
121              NaN  
122              NaN  
123              NaN  

[124 rows x 106 columns]

Then, we will get the precipitation data from the CSV file.

In [28]:
prcp = df[['DATE', 'PRCP']]
prcp

DATE   PRCP
0    1886  718.0
1    1887  642.1
2    1888  590.4
3    1889  875.8
4    1890  448.7
..    ...    ...
119  2018  860.3
120  2019  884.5
121  2020  664.1
122  2021  585.9
123  2022    NaN

[124 rows x 2 columns]

From 1886 to 2022, there are 137 years. However, result from the code above contains only 124 rows. If we do not consider the year 2022, there are 12 missed rows. Let's find them.

In [34]:
is_available = []
prcp_len = len(prcp)
# print(prcp_len)
beginning_year = int(prcp.iloc[0]['DATE'])
# print(beginning_year)
end_year = int(prcp.iloc[prcp_len - 1]['DATE'])
# print(end_year)
for i in range(beginning_year, end_year + 1, 1):
    is_available.append(False)

# print(len(is_available))

for i in range(prcp_len):
    if np.isnan(prcp.iloc[i]['PRCP']):
        continue
    is_available[int(prcp.iloc[i]['DATE']) - beginning_year] = True

missing_years = []
for i in range(len(is_available)):
    if is_available[i] == False:
        missing_years.append(i + beginning_year)

print(missing_years)
# len(missing_years)

[1949, 1950, 1951, 1952, 1953, 1954, 1955, 1956, 1957, 1958, 1959, 1960, 1961, 1962, 2022]


And now, we have 13 missing years from 1949 to 1961. Two years 1962 and 2022 have the NaN value.